### ISSUE TITLE
Shared contract between delphi server and client

### ISSUE THREAD

**Issue by radexpol** [danieleteti/delphimvcframework#614]
Does MVC supports something like contracts sharing, to use the same code on the server side and client side ? 

Pseudo Code:

**Server:**

```
type
  //shared contract 
 ISchool = interface
    [MVCPath('/GetTeachers')]
    procedure GetTeachers;
    procedure GetWorkingDays;  
    procedure SetSchoolName(const name: string);
  end; 
  
  [MVCPath('/api/school')]
  TSchoolController = class(TMVCController, ISchool)
    procedure GetTeachers;
    procedure GetWorkingDays;   
    procedure SetSchoolName(const name: string);
end;

```
**Client**
```
var schoolInterface: ISchool;

fRESTClient.Get('/api/school', schoolInterface);
 
var teachers: class(TList<TTeacher>);
teachers  := schoolInterface.GetTeachers;

```

**Comment by radexpol** [comment]
OK, I just made some experimental steps to get contracts working:

### Server
MVCFramework.Router.pas
replaced:
```

            if (LMethod.MethodKind <> mkProcedure) { or LMethod.IsClassMethod } then
              Continue;

```
with:
```
            if not (LMethod.MethodKind in [mkProcedure, mkFunction]) { or LMethod.IsClassMethod } then
              Continue;
```
MVCFramework.pas
replaced
```
                    try
                   lRouter.MethodToCall.Invoke(lSelectedController, lActualParams);
                    finally
                      lSelectedController.OnAfterAction(lContext, lRouterMethodToCallName);
                    end;

```
with:
```
                    try
                      if lRouter.MethodToCall.MethodKind = mkProcedure then
                        lRouter.MethodToCall.Invoke(lSelectedController, lActualParams)
                      else
                      begin
                        lResponseObject := lRouter.MethodToCall.Invoke(lSelectedController, lActualParams).AsObject;
                        lSelectedController.Render(lResponseObject);
                      end;
                    finally
                      lSelectedController.OnAfterAction(lContext, lRouterMethodToCallName);
                    end;

```

### controller
```
type  //shared unit
  TSum = class
  private
    FValue: integer;
  public
    property Value: integer read FValue write FValue;
  end;

  [MVCPath('/')]
  TApp1MainController = class(TMVCController)
    [MVCPath('/add')]
    [MVCHTTPMethod([httpGET])]
    function Add: TSum;

implementation

function TApp1MainController.Add: TSum;
begin
  Result := TSum.Create;
  Result.Value := 12344;
end;

```

### client

```
var
  Clt: IMVCRESTClient;
begin
  Clt := TMVCRESTClient.New.BaseURL('http://localhost', 8080);
  ShowMessage(Clt.Get('/add').Content); 
```
**-> response is {"Value":12344}**

this is just an experiment but I think that common contract would be a great feature for delphi MVC users

**Comment by danieleteti** [comment]
Nice POV, but did you already checked the RESTAdapter?
samples\wincellarclient_withrestadapter\

**Comment by danieleteti** [comment]
Also, IMHO REST APIs are really flexible. In some cases you need to render diofferent objects or inject link to create a proper HATEOAS. For something like you propose, generally can be used JSONRPC APIs and DMVCFramework offer a very strong support for JSONRPC (check sample named "jsonrpc_with_published_objects"). 
What do you think about? Do you still see some benefits in this different approach?
If you need to share models between server and client, what's the problem in the current scenario?

**Comment by radexpol** [comment]
Jsonrpc does not support swagger which is the big benefit of Rest when you must share your API to 3rd party companies/teams

**Comment by radexpol** [comment]
> Nice POV, but did you already checked the RESTAdapter? samples\wincellarclient_withrestadapter\

This sample does not share contract (common unit/s with endpoint definition and possible classes/lists/enums used in methods). Server creates the response in his own method, the client reads the data and put into his own object. That's not the way for reliable data exchange. I came from mORMot world where the contracts was the most important data exchange mechanism, and client contract checksum was verified with server checksum. That's why I was surprised that delphimvcframework does not contain such mechanism. But in opposite to mORMot, delphimvcframework is extremally nice and clearly written, your book (I bought) allows to enter the REST world with swagger, tokens, etc very, very fast.

My proposal was an extension for mvc and the existing code should work (procedures). In my opinion, hiding the raw endpoints from delphi enduser gives the opportunity to promote the methods of interfaces over necessity manually requesting for json string and passing query params to RestClient. The invisible part of this interface should make the request to backend.

**Comment by danieleteti** [comment]
It makes sense. I think we can do this change in the next beta version. 
Within next days will be released the 3.2.3-radium.
Then we can work on this.

**Comment by danieleteti** [comment]
DMVCFramework is really oriented to the "outside world" so it makes no assumption about the "client side". It can be Delphi, Javascript, Python or whatever so the "client contract" is not a thing in dmvcframework. However, give the possibile to implement actions as functions and not only as procedures is a recurrent topic in "my own roadmap". So, if it can be useful also for other things (also for delphi clients) why not?
This will be implemented in the `dmvcframework-3.3.0-fluorine` (the next beta).
